File: binary_tree_from_pre_in.cpp

// Time Complexity: O(N)
//Space Complexity: O(N)


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        int p = 0;
     
        return buildTreeHelper(preorder,inorder,0,inorder.size()-1,p);
        
    }

    
public:
    TreeNode* buildTreeHelper(vector<int>& preorder, vector<int>& inorder,int is,int ie,int &preIterator){
        // static int preIterator = 0;
        if(is>ie)return NULL;
        int pos;
        
        TreeNode* temp = new TreeNode(preorder[preIterator++]);
        if(is == ie)return temp;
        for(int i = is; i<=ie;i++){
            if(inorder[i]==temp->val){
                pos = i;
                break;
            }
        }
        temp->left = buildTreeHelper(preorder,inorder,is,pos-1,preIterator);
        temp->right = buildTreeHelper(preorder,inorder,pos+1,ie,preIterator);


        return temp;
    }
};

File: BinaryTreeFromPreIn.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 /**
 This Java solution constructs a binary tree from preorder and inorder traversal arrays using a recursive approach:
The buildTree function initializes the process and calls the recursive helper function with appropriate parameters.
The helper function creates the root node from the current preorder value, finds its index in inorder array to determine left and right subtrees, and recursively builds the left and right subtrees.
The process continues until the preidx reaches the end of the preorder array or the left index surpasses the right index.
Time Complexity: O(N) - where N is the number of nodes in the binary tree, as each node is processed once.
Space Complexity: O(H) - where H is the height of the binary tree. The space used by the recursive call stack is proportional to the height of the tree. 
In the worst case, the height is O(N), making the space complexity O(N).


  */
  class Solution {
    private int preidx = 0;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return helper(preorder, inorder, 0, preorder.length-1);
        
        
    }
    private TreeNode helper(int[] preorder, int[] inorder, int left, int right){
        if(preidx == preorder.length || left>right) return null;
        TreeNode root = new TreeNode(preorder[preidx]);
        for(int i=left;i<=right;i++){
            if(preorder[preidx] == inorder[i]){
                preidx++;
                root.left = helper(preorder, inorder, left, i-1);
                root.right = helper(preorder, inorder,  i+1, right);
                break;
            }

        }
        return root;
    }
}